tests/braille-specs/tr.yaml

```diff
@@ -0,0 +1,870 @@
+# Test file for Turkish 8 dots table, Turkish grade 1 table and Turkish grade 2 table
+#
+# Written by Attila Hammer and proofread by Oğuz Uğur
+#
+# Copyright (C) 2024 Attila Hammer from IT Foundation for the Visually Impaired - Hungary <www.infoalap.hu>
+# Copyright (C) 2024 Oğuz Uğur <oguzugur2834@gmail.com>
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved. This file is offered as-is,
+# without any warranty.
+
+display: unicode-without-blank.dis
+
+# ================
+# Computer braille
+# ================
+
+table:
+  language: tr
+  type: computer
+  __assert-match: tr.tbl
+flags: {testmode: forward}
```

Comment:
tr.tbl contains `#+direction:both` (it was marked this way because it is a computer braille table), however there areno tests for backward translation. If I change this line to `{testmode: bothDirections}`, a lot a tests fail.

Reply:
My question was mainly about the computer braille table, because usually computer braille tables provide a one-to-one mapping.

I think I would indeed prefer the tests that work in both directions, to run in both direction (only for the computer braille table). The ones that can't be back-translated because of conflicts could get a `{xfail: {backward: conflict}}` flag.

Reply:
Thanks Bert, ok, I think I understanding what you would like.
I will looking the computer braille entries the test case file and marks the conflicting parts the test case file with you wrote flag.
My question:
the three equals character into the test file are need? or an unwanted change before or after the table header part (I am not remember the proper place, but very similar a perhaps merge conflicting header part, when a conflict part marks three equals character an already putted part a file).
Another change I need doing before merge operation?

Attila

Reply:
I don't understand the question.

Reply:
I already doed the bothDirections change with computer Braille related testcase part.
Now I doing marking the braille dots conflicting back-translation failures with the {xfail: {backward: conflict}} flag.
Previous, when I created the testcase files I forgot lou_checkyaml already supports these failure markups, sorry, next time I using this technique.
Lot of conflicts are impossible resolve with table level, because the \x00e0, \x00e1 or \x00e2 characters are equals need marking I think with dots 4-1 braille dots.

Bert, I doing similar bothDirections mode change with grade1 and grade 2 parts the testcase, or enough the computer braille part into the test file?
I would like doing only one commit and one push the test file related. :-):-)

Attila

Reply:
Bert, I am suspicious turkish grade1 and grade2 tables is not back-translation compatible yet.
When I change forward with bothDirections mode, I get 200 failures with back-translation direction only the grade 1 table related, not all failures is dot conflicts, but very lot of dots are conflicting another braille equals dots, similar with computer Braille table.
Looks for example the uppercase A letter back-translation result, very interesting the result, because I am not known where from apply Liblouis a pass4 rule or an another rule with produce wrong output, only sure to I am not putted any pass2 rules the turkish grade 1 table. With a code part I putting lou_trace -b command result with unicode.dis,tr-g1.ctb command result:
```
tools/lou_trace unicode.dis,tr-g1.ctb -b
⠠⠁
◌a
1.	pass2	$sldpUumSDwxyz	*#1=0
2.	pass2	$d	*
3.	always	◌	6
4.	lowercase	a	1
```
What? Why? :-):-)
Turkish grade 1 table doesn't contains pass2 rules, perhaps this pass2 rules coming from en-ueb-g1.ctb or an another subtable with en-ueb-g1.ctb table are included.
The \x25cc character is need marking with dot 6,
I am not known visual the \x25cc character is a turkish letter?
Perhaps if we putting a noback prefix before with \x25cc character rule, possible resolve the situation with uppercase a related, but this is only one failure.  :-):-)
So:
The computer Braille related I marked back-translation conflicting test cases with you requested flag the testcase entries, and in computer braille section I changed the forward direction with bothDirections flag.
Grade1 test section I keeped the forward direction, because if I changing the direction with bothDirections mode, I get 200 failure results.
I not looked the grade2 results, but I am suspicious the failure result quality perhaps similar.
I suggesting because have only I think four day before release publication, now keeps forward direction with grade1 and grade2 testcases, and trying solve grade1 testcase bothDirections change the 3.30 development cicle, because the failure quality require more debugging work with back-translation direction.
You accept this compromise suggestion?
If yes, in the 3.30 cicle I trying debugging turkish grade 1 table and trying decrease the back-translation failure quality. :-):-)
If not accepts this compromise suggestion, I trying too debugging the 3.30 cicle the turkish grade1 table, because disturbing me the 200 failures with back-translation direction. :-):-)
This situation need marking in the test file the grade 1 section what test are conflicting xfail testcase, and what test are simple xfailing test with need fixing the future. :-):-)

Attila

Reply:
OK that is fine.

Reply:
This is interesting...
I examined the grade 2 bothDirection mode, the tests produces only 163 failures to back-translation direction, similar with computer Braille tests.
My original grade2 test file contains 286 test entries, this is not a bad percentage. :-):-)
So, in the 3.30 development cicle absolute sure need improve tr-g1.ctb table back-translation quality, because my original grade 1 file contains 255 test entries, and 200 entries are back-translation failing. This is horrible quality for back-translation direction, forward direction is fine.
Perhaps the 3.30 cicle entire need redesigning tr-g1.ctb table, and need remove the last en-ueb-g1.ctb include line.
But I don't understand a think, perhaps future need consulting Arend Amend, he's the original table author I think in 2017 if I remember correct from commit history:
Original table creation time Arend why designed the tr-g1.ctb table to include en-ueb-g1.ctb table? Why very need this include line?
Only latin letters not require this include line, sure have purpose this, not an unvanted include operation happened by design (Arend Amend I think an experienced table author).
I see already complicated complex Braille systems, similar with hungarian language (hungarian language braille systems is relative complex, turkish braille is absolute more complex), but the turkish braille system is an interesting „mix” for braille related (contains english braille systems rules, but contains german and french braille systems based rules, and few irish language based rules). :-):-)
I understanded this from this very fresh Wikipedia article when I worked table fixes and want little learning with turkish braille to I understanding what and why need doing a required way:
[Turkish Braille Wikipedia Article](https://en.wikipedia.org/wiki/Turkish_Braille)

So:
Now I marking conflicting back-translation entries with the proper flag the grade 2 test case entries, similar with computer braille tests already we doed, and change grade2 test direction with bothDirections flags.
Perhaps now we keeping the grade 1 tests with forward direction, and after 3.29.0 release publication we redesigning the grade1 table, together with Arend Amend if he subscribed the Liblouis list.
Forward direction presenting is fine, need improve the back-translation quality if this is possible, except the equals dot combinations related rules (impossible to fixing this type rules with back-translation direction, because this rules uses equals dot combinations).

Attila

Reply:
Oh... I am possible was little tired when I added the \x25cc symbol the tables/tr-g1.ctb table, simple I forgotted putting the noback prefix before always opcode (now detected this). :-):-)
After I doed this change, the 200 failures in grade 1 both Directions mode decreased with 136 failures, and the 136 failures are usual with braille dots conflicting failures, similar with other turkish 8 dots and grade2 tables).
Now I marks conflicting failures the grade 1 tests the proper flags, and do a final push when I am ready.
When I worked the grade 1 table update, @oguzugur28 sent me a large character and correct list (I think the list have more than 141 rules), when comed the evening, I seed all place with unicode hexa codes with all places. :-):-)
So, the final conclusion:
* All tables, all test case parts are bothDirections mode,
* All test sections I will be marking with conflicting xfail entries. Unfortunatelly with future this test case entries impossible to resolve back-translation direction the failures, because all characters uses equals dot combinations, bot forward and backward direction. With turkish language if I known right not have special braille input prefix, for example with dot 5 when a dot combination need doing different meaning with back-translation direction (a typical example the hungarian 1456-1456, and 1456-5-1456 with back-translation direction. If back-translation the hungarian user type two 146 dot combination he's perkins keyboard, need back-translate ccs text, if type for example 1456-5-1456, need back-translate cscs text). With turkish language not awailable this type rules and the standard are not defining this type rules.

Boys, I think we doed everiting with turkish computer braille, grade 1 braille and grade 2 braille with this is possible. The final result will be not too bad. So, in finish, have the happy end! :-):-)
So, I am finishing the grade 1 test failures markup, and do a final push.
After this, the new committed changes are review ready I think.
Sorry the previous not detected forgotted noback prefix, this is already fixed with tables/tr-g1.ctb, but not pushed yet.

Attila

Reply:
Ok, I changed grade 1 tests with bothDirections mode, and marked the conflicting entries with following flag:
{xfail: {backward: conflict}}
I remember wrong number with grade 1 tests, from 255 or 256 tests after the noback related previous change only 133 tests are failed.
So, this development cicle I done the turkish table related work, entire yesterday and this day added changes are review ready my side.
Both three Bert doed test case sections the tests/braille-specs/tr.yaml file are bothDirections mode, all test sections are marked the conflicting back-translation xfail entryes.

I wish everybody a good weekend and good rest if this is possible, tomorrow I will be limited near by internet, sunday entire day I am not in home.

Everybody I wishing a good 3.29.0 Liblouis release in next week monday. :-):-)

Thanks the cooperation during review,

Attila

Reply:
The third conversion part with test files related are possible resolve, all tasks done with test files related.
I will quote other comment part from @bertfrees the proper sections.
Usual this parts are technical questions, with me unable to answering, because native I am not known turkish language.
So, if everibody accepts this, I am or perhaps Chris or Bert resolve the test files related conversion part to easying see remaining questions and tasks.